bounded/src/set.rs

```diff
@@ -128,132 +123,156 @@ pub fn check(
     }
 }
 
 /// A BoundedProgram is a set of initial states, a set of transitions, and a safety property
 #[derive(Clone, Debug, PartialEq)]
-pub struct BoundedProgram {
+struct BoundedProgram {
     /// List of initial states
     inits: Vec<BoundedState>,
     /// List of transitions to potentially take at each step. The transition relation is the
     /// disjunction of all these transitions.
     trs: Vec<Transition>,
-    /// Safety property to check in each reachable state, expressed in DNF, i.e., the outer Vec is
-    /// an "or" of the inner Vecs, and each inner Vec is the "and" of its Guards.
-    safes: Vec<Vec<Guard>>,
+    /// Safety property to check in each reachable state.
+    safe: Formula,
 }
 
 /// A Transition is a deterministic partial function on states expressed as a guarded update.
 /// If the (and of the) guard(s) is true, then the transition is enabled and can step to the updated
 /// state.
 /// If the (and of the) guard(s) is false, then the transition is not enabled.
 #[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct Transition {
+struct Transition {
     guards: Vec<Guard>,
     updates: Vec<Update>,
-}
-
-impl Transition {
-    /// Make a Transition
-    pub fn new(guards: Vec<Guard>, updates: Vec<Update>) -> Transition {
-        Transition { guards, updates }
-    }
+    slow_guard: Formula,
 }
 
 /// A Guard is a logical literal, i.e., a possibly negated relation applied to an argument tuple
 /// such as `r(x, y)` or `!r(x, y)`. The relation and argument tuple are represented by an index
-/// into an ambient `Indices` map.
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct Guard {
-    /// The index representing the relation and argument tuple. Indexes into an ambient `Indices`
+/// into an ambient `Context` map.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+struct Guard {
+    /// The index representing the relation and argument tuple. Indexes into an ambient `Context`
     /// map.
     index: usize,
-
     /// True for positive literal, false for negative literal
     value: bool,
 }
 
 /// An Update is either an insertion or a removal of a tuple from a relation. The relation and
-/// argument tuple are represented by an index into an ambient `Indices` map.
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct Update {
-    /// The index representing the relation and argument tuple. Indexes into an ambient `Indices`
+/// argument tuple are represented by an index into an ambient `Context` map.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+struct Update {
+    /// The index representing the relation and argument tuple. Indexes into an ambient `Context`
     /// map.
     index: usize,
-
     /// True for insertion, false for removal
-    value: bool,
+    formula: Formula,
+}
+
+impl Transition {
+    fn from_conjunction(trs: impl IntoIterator<Item = Transition>) -> Transition {
```

Comment:
Document that this is "checked-parallel" composition.

Would it make sense to move the assertion that guards/updates don't overlap into this function?